api/v1alpha1/workflow_webhook.go

```diff
@@ -142,35 +142,31 @@ func (w *Workflow) ValidateUpdate(old runtime.Object) error {
 		}
 	}
 
-	// New state is the desired state in the Spec
-	newState, err := GetWorkflowState(w.Spec.DesiredState)
-	if err != nil {
-		return fmt.Errorf("unable to parse DesiredState: %w", err)
-	}
+	oldState := oldWorkflow.Status.State
+	newState := w.Spec.DesiredState
 
-	// Old state is the current state we're on. This is found in the status
-	oldState, err := GetWorkflowState(oldWorkflow.Status.State)
-	if err != nil {
-		return fmt.Errorf("unable to find current state: %w", err)
+	// Allow updates without changes to state, or to the next state. Teardown is always permitted
+	if newState == oldState || oldState.next() != newState || newState == StateTeardown {
+		return nil
 	}
 
-	// Progressing to teardown is allowed at any time, and changes to the
-	// Workflow that don't change the state are fine too (immutable fields were
-	// already checked)
-	if newState == StateTeardown || newState == oldState {
-		return nil
+	if !oldWorkflow.Status.Ready {
+		return fmt.Errorf("current desired state not yet achieved")
 	}
 
-	if newState < oldState {
+	// Error checks
+	// TODO: These (and other errors) should move to validation field errors. Refer to
+	//       https://pkg.go.dev/k8s.io/apimachinery@v0.24.2/pkg/util/validation/field
+	if oldState.after(newState) {
 		return fmt.Errorf("state cannot progress backwards")
 	}
 
-	if newState > oldState+1 {
+	if oldState.next() != newState {
 		return fmt.Errorf("states cannot be skipped")
 	}
 
-	if !oldWorkflow.Status.Ready {
-		return fmt.Errorf("current desired state not yet achieved")
+	if oldState.last() {
```

Comment:
Not sure this can happen, given that kubebuilder has validated the string at this point and there are no strings beyond "Teardown".

Reply:
A little bonus check. I removed it